File: genium/src/main/java/com/here/genium/generator/cpp/CppModelBuilder.java
/*
 * Copyright (c) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.ivi.api.generator.cpp;

import com.google.common.annotations.VisibleForTesting;
import com.here.ivi.api.common.CollectionsHelper;
import com.here.ivi.api.common.FrancaTypeHelper;
import com.here.ivi.api.generator.common.modelbuilder.AbstractModelBuilder;
import com.here.ivi.api.generator.common.modelbuilder.ModelBuilderContextStack;
import com.here.ivi.api.model.common.InstanceRules;
import com.here.ivi.api.model.cpp.*;
import com.here.ivi.api.model.franca.CommentHelper;
import com.here.ivi.api.model.franca.FrancaDeploymentModel;
import java.util.*;
import org.franca.core.franca.*;

public class CppModelBuilder extends AbstractModelBuilder<CppElement> {

  private final FrancaDeploymentModel deploymentModel;
  private final CppTypeMapper typeMapper;

  @VisibleForTesting
  CppModelBuilder(
      final ModelBuilderContextStack<CppElement> contextStack,
      final FrancaDeploymentModel deploymentModel,
      final CppTypeMapper typeMapper) {
    super(contextStack);
    this.deploymentModel = deploymentModel;
    this.typeMapper = typeMapper;
  }

  public CppModelBuilder(
      final FrancaDeploymentModel deploymentModel, final CppTypeMapper typeMapper) {
    this(new ModelBuilderContextStack<>(), deploymentModel, typeMapper);
  }

  @Override
  public void finishBuilding(FInterface francaInterface) {

    String className = CppNameRules.getClassName(francaInterface.getName());
    String classComment = CommentHelper.getDescription(francaInterface);

    CppClass cppClass = new CppClass(className, classComment);

    FInterface parentInterface = francaInterface.getBase();
    if (parentInterface != null) {
      CppTypeRef parentTypeRef = typeMapper.mapComplexType(parentInterface);
      cppClass.inheritances.add(new CppInheritance(parentTypeRef, CppInheritance.Type.Public));
    }

    cppClass.methods.addAll(getPreviousResults(CppMethod.class));
    cppClass.members.addAll(getPreviousResults(CppEnum.class));
    cppClass.members.addAll(getPreviousResults(CppUsing.class));
    cppClass.members.addAll(getPreviousResults(CppStruct.class));

    storeResult(cppClass);
    closeContext();
  }

  @Override
  public void finishBuilding(FMethod francaMethod) {

    CppParameter outputParameter =
        CollectionsHelper.getStreamOfType(getCurrentContext().previousResults, CppParameter.class)
            .filter(parameter -> parameter.isOutput)
            .findFirst()
            .orElse(null);
    String returnTypeComment = outputParameter != null ? outputParameter.comment : null;

    CppTypeRef errorEnumTypeRef = getPreviousResult(CppTypeRef.class);
    CppTypeRef errorType = errorEnumTypeRef != null ? CppTypeMapper.STD_ERROR_CODE_TYPE : null;
    CppTypeRef returnType = mapMethodReturnType(outputParameter, errorType);
    CppMethod cppMethod = buildCppMethod(francaMethod, returnType, returnTypeComment);

    storeResult(cppMethod);
    closeContext();
  }

  @Override
  public void finishBuildingInputArgument(FArgument francaArgument) {

    CppTypeRef cppTypeRef = getPreviousResult(CppTypeRef.class);
    CppParameter cppParameter =
        new CppParameter(CppNameRules.getParameterName(francaArgument.getName()), cppTypeRef);
    cppParameter.comment = CommentHelper.getDescription(francaArgument);

    storeResult(cppParameter);
    closeContext();
  }

  @Override
  public void finishBuildingOutputArgument(FArgument francaArgument) {

    CppTypeRef cppTypeRef = getPreviousResult(CppTypeRef.class);
    CppParameter cppParameter =
        new CppParameter(CppNameRules.getParameterName(francaArgument.getName()), cppTypeRef, true);
    cppParameter.comment = CommentHelper.getDescription(francaArgument);

    storeResult(cppParameter);
    closeContext();
  }

  @Override
  public void finishBuilding(FTypeCollection francaTypeCollection) {

    for (CppElement cppElement : getCurrentContext().previousResults) {
      if (cppElement instanceof CppEnum
          || cppElement instanceof CppConstant
          || cppElement instanceof CppUsing
          || cppElement instanceof CppStruct) {
        storeResult(cppElement);
      }
    }

    closeContext();
  }

  @Override
  public void finishBuilding(FConstantDef francaConstant) {

    CppTypeRef cppTypeRef = getPreviousResult(CppTypeRef.class);
    CppValue value = CppValueMapper.map(francaConstant.getRhs());

    String name = CppNameRules.getConstantName(francaConstant.getName());
    String fullyQualifiedName = CppNameRules.getConstantFullyQualifiedName(francaConstant);
    CppConstant cppConstant = new CppConstant(name, fullyQualifiedName, cppTypeRef, value);
    cppConstant.comment = CommentHelper.getDescription(francaConstant);

    storeResult(cppConstant);
    closeContext();
  }

  @Override
  public void finishBuilding(FField francaField) {

    CppTypeRef cppTypeRef = getPreviousResult(CppTypeRef.class);
    String fieldName = CppNameRules.getFieldName(francaField.getName());

    String deploymentDefaultValue = deploymentModel.getDefaultValue(francaField);
    CppValue cppValue =
        deploymentDefaultValue != null
            ? CppValueMapper.mapDeploymentDefaultValue(cppTypeRef, deploymentDefaultValue)
            : null;

    CppField cppField =
        CppField.builder(fieldName, cppTypeRef)
            .initializer(cppValue)
            .isNotNull(deploymentModel.isNotNull(francaField))
            .build();
    cppField.comment = CommentHelper.getDescription(francaField);

    storeResult(cppField);
    closeContext();
  }

  @Override
  public void finishBuilding(FStructType francaStructType) {

    // Type definition

    String name = CppNameRules.getStructName(francaStructType.getName());
    String fullyQualifiedName = CppNameRules.getFullyQualifiedName(francaStructType);
    String comment = CommentHelper.getDescription(francaStructType);
    CppStruct cppStruct = new CppStruct(name, fullyQualifiedName, comment);

    cppStruct.fields.addAll(getPreviousResults(CppField.class));

    CppTypeRef parentTypeRef = getPreviousResult(CppTypeRef.class);
    if (parentTypeRef != null) {
      cppStruct.inheritances.add(new CppInheritance(parentTypeRef, CppInheritance.Type.Public));
    }
    storeResult(cppStruct);

    // Type reference
    storeResult(typeMapper.mapComplexType(francaStructType));

    closeContext();
  }

  @Override
  public void finishBuilding(FTypeDef francaTypeDef) {

    if (!InstanceRules.isInstanceId(francaTypeDef)) {
      String typeDefName = CppNameRules.getTypedefName(francaTypeDef.getName());
      String fullyQualifiedName = CppNameRules.getFullyQualifiedName(francaTypeDef);
      CppTypeRef cppTypeRef = getPreviousResult(CppTypeRef.class);
      String comment = CommentHelper.getDescription(francaTypeDef);
      CppUsing cppUsing =
          CppUsing.builder(typeDefName, cppTypeRef)
              .fullyQualifiedName(fullyQualifiedName)
              .comment(comment)
              .build();

      storeResult(cppUsing);
    }

    closeContext();
  }

  @Override
  public void finishBuilding(FArrayType francaArrayType) {

    String name = CppNameRules.getTypedefName(francaArrayType.getName());
    String fullyQualifiedName = CppNameRules.getFullyQualifiedName(francaArrayType);
    CppTypeRef elementType = getPreviousResult(CppTypeRef.class);
    CppTypeRef targetType =
        CppTemplateTypeRef.create(CppTemplateTypeRef.TemplateClass.VECTOR, elementType);
    String comment = CommentHelper.getDescription(francaArrayType);
    CppUsing cppUsing =
        CppUsing.builder(name, targetType)
            .fullyQualifiedName(fullyQualifiedName)
            .comment(comment)
            .build();

    storeResult(cppUsing);
    closeContext();
  }

  @Override
  public void finishBuilding(FMapType francaMapType) {

    String name = CppNameRules.getTypedefName(francaMapType.getName());
    String fullyQualifiedName = CppNameRules.getFullyQualifiedName(francaMapType);
    List<CppTypeRef> typeRefs = getPreviousResults(CppTypeRef.class);
    CppTypeRef targetType = typeMapper.wrapMap(typeRefs.get(0), typeRefs.get(1));
    String comment = CommentHelper.getDescription(francaMapType);
    CppUsing cppUsing =
        CppUsing.builder(name, targetType)
            .fullyQualifiedName(fullyQualifiedName)
            .comment(comment)
            .build();

    storeResult(cppUsing);
    closeContext();
  }

  @Override
  public void finishBuilding(FEnumerationType francaEnumerationType) {

    // Type definition
    String enumName = CppNameRules.getEnumName(francaEnumerationType.getName());
    String fullyQualifiedName = CppNameRules.getFullyQualifiedName(francaEnumerationType);
    CppEnum cppEnum = CppEnum.createScoped(enumName, fullyQualifiedName);
    cppEnum.comment = CommentHelper.getDescription(francaEnumerationType);
    cppEnum.items.addAll(getPreviousResults(CppEnumItem.class));
    storeResult(cppEnum);

    // Type reference
    storeResult(typeMapper.mapEnum(francaEnumerationType));

    closeContext();
  }

  @Override
  public void finishBuilding(FEnumerator francaEnumerator) {

    String enumItemName = CppNameRules.getEnumEntryName(francaEnumerator.getName());
    CppValue cppValue = getPreviousResult(CppValue.class);
    CppEnumItem cppEnumItem = new CppEnumItem(enumItemName, cppValue);
    cppEnumItem.comment = CommentHelper.getDescription(francaEnumerator);

    storeResult(cppEnumItem);
    closeContext();
  }

  @Override
  public void finishBuilding(FExpression francaExpression) {

    storeResult(CppValueMapper.map(francaExpression));
    closeContext();
  }

  @Override
  public void finishBuilding(FTypeRef francaTypeRef) {

    CppTypeRef cppTypeRef = typeMapper.map(francaTypeRef);

    if (FrancaTypeHelper.isImplicitArray(francaTypeRef)) {
      cppTypeRef = CppTemplateTypeRef.create(CppTemplateTypeRef.TemplateClass.VECTOR, cppTypeRef);
    }

    storeResult(cppTypeRef);
    closeContext();
  }

  @Override
  public void finishBuilding(FAttribute francaAttribute) {

    CppTypeRef cppTypeRef = getPreviousResult(CppTypeRef.class);
    String francaComment = CommentHelper.getDescription(francaAttribute);

    String getterName = CppNameRules.getGetterName(francaAttribute.getName(), cppTypeRef);
    CppMethod getterMethod = buildAccessorMethod(getterName, cppTypeRef, francaComment);
    storeResult(getterMethod);

    if (!francaAttribute.isReadonly()) {
      String setterName = CppNameRules.getSetterName(francaAttribute.getName());
      CppMethod setterMethod =
          buildAccessorMethod(setterName, CppPrimitiveTypeRef.VOID, francaComment);
      setterMethod.parameters.add(new CppParameter("value", cppTypeRef));
      storeResult(setterMethod);
    }

    closeContext();
  }

  private CppMethod buildAccessorMethod(
      final String name, final CppTypeRef cppTypeRef, final String francaComment) {

    return new CppMethod.Builder(name)
        .returnType(cppTypeRef)
        .specifier(CppMethod.Specifier.VIRTUAL)
        .qualifier(CppMethod.Qualifier.PURE_VIRTUAL)
        .comment(francaComment)
        .build();
  }

  private CppMethod buildCppMethod(
      final FMethod francaMethod, final CppTypeRef returnType, final String returnComment) {

    String methodName = CppNameRules.getMethodName(francaMethod.getName());
    String fullyQualifiedMethodName =
        CppNameRules.getFullyQualifiedName(
            CppNameRules.getNestedNameSpecifier(francaMethod), methodName);
    CppMethod.Builder builder =
        new CppMethod.Builder(methodName)
            .fullyQualifiedName(fullyQualifiedMethodName)
            .returnType(returnType);

    if (deploymentModel.isStatic(francaMethod)) {
      builder.specifier(CppMethod.Specifier.STATIC);
    } else {
      if (deploymentModel.isConst(francaMethod)) {
        // const needs to be before "= 0" pure virtual specifier
        builder.qualifier(CppMethod.Qualifier.CONST);
      }
      builder.specifier(CppMethod.Specifier.VIRTUAL);
      builder.qualifier(CppMethod.Qualifier.PURE_VIRTUAL);
    }

    builder.comment(CommentHelper.getDescription(francaMethod));
    builder.returnComment(returnComment);

    CollectionsHelper.getStreamOfType(getCurrentContext().previousResults, CppParameter.class)
        .filter(parameter -> !parameter.isOutput)
        .forEach(builder::parameter);

    return builder.build();
  }

  private CppTypeRef mapMethodReturnType(
      final CppParameter outputParameter, final CppTypeRef errorType) {

    CppTypeRef outArgType = outputParameter != null ? outputParameter.type : null;
    if (errorType == null && outArgType == null) {
      return CppPrimitiveTypeRef.VOID;
    } else if (errorType != null && outArgType == null) {
      return errorType;
    } else if (errorType == null) {
      return outArgType;
    }

    // wrap multiple out values (error + outArg) in their own type
    return typeMapper.getReturnWrapperType(outArgType, errorType);
  }
}

File: genium/src/main/java/com/here/genium/generator/cpp/CppModelBuilder.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cpp

import com.google.common.annotations.VisibleForTesting
import com.here.genium.generator.common.modelbuilder.AbstractLimeBasedModelBuilder
import com.here.genium.common.ModelBuilderContextStack
import com.here.genium.model.common.CommentsPreprocessor
import com.here.genium.model.cpp.CppClass
import com.here.genium.model.cpp.CppConstant
import com.here.genium.model.cpp.CppElement
import com.here.genium.model.cpp.CppEnum
import com.here.genium.model.cpp.CppEnumItem
import com.here.genium.model.cpp.CppField
import com.here.genium.model.cpp.CppInheritance
import com.here.genium.model.cpp.CppMethod
import com.here.genium.model.cpp.CppParameter
import com.here.genium.model.cpp.CppPrimitiveTypeRef
import com.here.genium.model.cpp.CppStruct
import com.here.genium.model.cpp.CppTypeRef
import com.here.genium.model.cpp.CppUsing
import com.here.genium.model.cpp.CppValue
import com.here.genium.model.lime.LimeAttributeType
import com.here.genium.model.lime.LimeBasicType
import com.here.genium.model.lime.LimeConstant
import com.here.genium.model.lime.LimeContainer
import com.here.genium.model.lime.LimeEnumeration
import com.here.genium.model.lime.LimeEnumerator
import com.here.genium.model.lime.LimeField
import com.here.genium.model.lime.LimeMethod
import com.here.genium.model.lime.LimeParameter
import com.here.genium.model.lime.LimeProperty
import com.here.genium.model.lime.LimeStruct
import com.here.genium.model.lime.LimeTypeDef
import com.here.genium.model.lime.LimeTypeHelper
import com.here.genium.model.lime.LimeTypeRef
import com.here.genium.model.lime.LimeValue
import com.here.genium.model.lime.LimeValue.Special.ValueId
import java.util.EnumSet

class CppModelBuilder @VisibleForTesting internal constructor(
    contextStack: ModelBuilderContextStack<CppElement>,
    private val typeMapper: CppTypeMapper,
    private val nameResolver: CppNameResolver
) : AbstractLimeBasedModelBuilder<CppElement>(contextStack) {

    constructor(
        typeMapper: CppTypeMapper,
        nameResolver: CppNameResolver
    ) : this(ModelBuilderContextStack<CppElement>(), typeMapper, nameResolver)

    override fun finishBuilding(limeContainer: LimeContainer) {
        val members: List<CppElement> = getPreviousResults(CppEnum::class.java) +
            getPreviousResults(CppUsing::class.java) +
            getPreviousResults(CppStruct::class.java) +
            getPreviousResults(CppConstant::class.java)

        if (limeContainer.type == LimeContainer.ContainerType.TYPE_COLLECTION) {
            members.forEach { storeResult(it) }
            closeContext()
            return
        }

        val limeParentType = limeContainer.parent
        val inheritances = when {
            limeParentType != null -> {
                val parentType =
                    typeMapper.mapInstanceType(limeParentType.type as LimeContainer, false)
                listOf(CppInheritance(parentType, CppInheritance.Type.Public))
            }
            else -> emptyList()
        }
        val cppClass = CppClass(
            name = nameResolver.getName(limeContainer),
            fullyQualifiedName = nameResolver.getFullyQualifiedName(limeContainer),
            comment = limeContainer.comment,
            isExternal = limeContainer.attributes.have(LimeAttributeType.EXTERNAL_TYPE),
            members = members,
            methods = getPreviousResults(CppMethod::class.java),
            inheritances = inheritances,
            isEquatable = limeContainer.attributes.have(LimeAttributeType.EQUATABLE)
        )

        storeNamedResult(limeContainer, cppClass)
        closeContext()
    }

    override fun finishBuilding(limeMethod: LimeMethod) {
        val specifiers = EnumSet.noneOf(CppMethod.Specifier::class.java)
        val qualifiers = EnumSet.noneOf(CppMethod.Qualifier::class.java)
        if (limeMethod.isStatic || limeMethod.attributes.have(LimeAttributeType.CONSTRUCTOR)) {
            specifiers.add(CppMethod.Specifier.STATIC)
        } else {
            if (limeMethod.attributes.have(LimeAttributeType.CONST)) {
                // const needs to be before "= 0" pure virtual specifier
                qualifiers.add(CppMethod.Qualifier.CONST)
            }
            specifiers.add(CppMethod.Specifier.VIRTUAL)
            qualifiers.add(CppMethod.Qualifier.PURE_VIRTUAL)
        }

        val isNullable = limeMethod.returnType.attributes.have(LimeAttributeType.NULLABLE)
        val isInstance = limeMethod.returnType.typeRef.type is LimeContainer
        var cppReturnType = typeMapper.mapType(limeMethod.returnType.typeRef)
        if (isNullable && !isInstance) {
            cppReturnType = typeMapper.createOptionalType(cppReturnType)
        }
        val errorEnum = getPreviousResultOrNull(CppTypeRef::class.java)
        val returnType = when {
            errorEnum != null && cppReturnType != CppPrimitiveTypeRef.VOID ->
                typeMapper.getReturnWrapperType(
                    cppReturnType,
                    CppTypeMapper.STD_ERROR_CODE_TYPE
                )
            errorEnum != null -> CppTypeMapper.STD_ERROR_CODE_TYPE
            else -> cppReturnType
        }

        val cppMethod = CppMethod(
            nameResolver.getName(limeMethod),
            nameResolver.getFullyQualifiedName(limeMethod),
            limeMethod.comment,
            returnType,
            limeMethod.returnType.comment,
            errorEnum?.fullyQualifiedName,
            isInstance && !isNullable,
            getPreviousResults(CppParameter::class.java),
            specifiers,
            qualifiers
        )

        storeNamedResult(limeMethod, cppMethod)
        closeContext()
    }

    override fun finishBuilding(limeParameter: LimeParameter) {
        val isNullable = limeParameter.attributes.have(LimeAttributeType.NULLABLE)
        val isInstance = limeParameter.typeRef.type is LimeContainer
        var cppTypeRef = getPreviousResult(CppTypeRef::class.java)
        if (isNullable && !isInstance) {
            cppTypeRef = typeMapper.createOptionalType(cppTypeRef)
        }

        val cppParameter = CppParameter(
            name = nameResolver.getName(limeParameter),
            type = cppTypeRef,
            isNotNull = isInstance && !isNullable
        )
        cppParameter.comment = limeParameter.comment

        storeResult(cppParameter)
        closeContext()
    }

    override fun finishBuilding(limeStruct: LimeStruct) {
        val methods = getPreviousResults(CppMethod::class.java).map {
            val specifiers = setOf(CppMethod.Specifier.STATIC) intersect it.specifiers
            val qualifiers = when {
                specifiers.contains(CppMethod.Specifier.STATIC) -> emptySet()
                else -> setOf(CppMethod.Qualifier.CONST)
            }
            it.copy(specifiers, qualifiers)
        }
        val cppStruct = CppStruct(
            nameResolver.getName(limeStruct),
            nameResolver.getFullyQualifiedName(limeStruct),
            limeStruct.comment,
            limeStruct.attributes.have(LimeAttributeType.EXTERNAL_TYPE),
            getPreviousResults(CppField::class.java),
            methods,
            getPreviousResults(CppConstant::class.java),
            limeStruct.attributes.have(LimeAttributeType.EQUATABLE),
            limeStruct.attributes.have(LimeAttributeType.IMMUTABLE)
        )

        storeNamedResult(limeStruct, cppStruct)
        closeContext()
    }

    override fun finishBuilding(limeField: LimeField) {
        var cppTypeRef = getPreviousResult(CppTypeRef::class.java)
        val isNullable = limeField.attributes.have(LimeAttributeType.NULLABLE)
        val isInstance = limeField.typeRef.type is LimeContainer
        if (isNullable && !isInstance) {
            cppTypeRef = typeMapper.createOptionalType(cppTypeRef)
        }

        val allTypes = LimeTypeHelper.getAllFieldTypes(limeField.typeRef.type)
        val hasImmutableType = allTypes.any { it.attributes.have(LimeAttributeType.IMMUTABLE) }

        val cppField = CppField(
            name = nameResolver.getName(limeField),
            fullyQualifiedName = nameResolver.getFullyQualifiedName(limeField),
            type = cppTypeRef,
            initializer = getPreviousResultOrNull(CppValue::class.java),
            isNotNull = isInstance && !isNullable,
            isNullable = isNullable,
            hasImmutableType = hasImmutableType,
            isClassEquatable = isInstance && limeField.typeRef.type.attributes.have(LimeAttributeType.EQUATABLE)
        )
        cppField.comment = limeField.comment

        storeNamedResult(limeField, cppField)
        closeContext()
    }

    override fun finishBuilding(limeTypeDef: LimeTypeDef) {
        val cppUsing = CppUsing(
            nameResolver.getName(limeTypeDef),
            nameResolver.getFullyQualifiedName(limeTypeDef),
            limeTypeDef.comment,
            getPreviousResult(CppTypeRef::class.java)
        )

        storeNamedResult(limeTypeDef, cppUsing)
        closeContext()
    }

    override fun finishBuilding(limeProperty: LimeProperty) {
        var cppTypeRef = getPreviousResult(CppTypeRef::class.java)
        val isNullable = limeProperty.attributes.have(LimeAttributeType.NULLABLE)
        val isInstance = limeProperty.typeRef.type is LimeContainer
        val isNotNull = !isNullable && isInstance
        if (isNullable && !isInstance) {
            cppTypeRef = typeMapper.createOptionalType(cppTypeRef)
        }

        val specifiers = when {
            limeProperty.isStatic -> EnumSet.of(CppMethod.Specifier.STATIC)
            else -> EnumSet.of(CppMethod.Specifier.VIRTUAL)
        }
        val getterQualifiers = when {
            limeProperty.isStatic -> EnumSet.noneOf(CppMethod.Qualifier::class.java)
            else -> EnumSet.of(CppMethod.Qualifier.CONST, CppMethod.Qualifier.PURE_VIRTUAL)
        }

        val getterMethod = CppMethod(
            name = nameResolver.getGetterName(limeProperty),
            fullyQualifiedName = nameResolver.getFullyQualifiedGetterName(limeProperty),
            comment = CommentsPreprocessor.preprocessGetterComment(limeProperty.comment),
            returnType = cppTypeRef,
            isNotNull = isNotNull,
            specifiers = specifiers,
            qualifiers = getterQualifiers
        )
        storeNamedResult(limeProperty, getterMethod)

        if (!limeProperty.isReadonly) {
            val setterParameter = CppParameter("value", cppTypeRef, isNotNull)
            val setterQualifiers = when {
                limeProperty.isStatic -> EnumSet.noneOf(CppMethod.Qualifier::class.java)
                else -> EnumSet.of(CppMethod.Qualifier.PURE_VIRTUAL)
            }
            val setterMethod = CppMethod(
                name = nameResolver.getSetterName(limeProperty),
                fullyQualifiedName = nameResolver.getFullyQualifiedSetterName(limeProperty),
                comment = CommentsPreprocessor.preprocessSetterComment(limeProperty.comment),
                parameters = listOf(setterParameter),
                specifiers = specifiers,
                qualifiers = setterQualifiers
            )
            storeResult(setterMethod)
        }

        closeContext()
    }

    override fun finishBuilding(limeEnumeration: LimeEnumeration) {
        val cppEnum = CppEnum(
            nameResolver.getName(limeEnumeration),
            nameResolver.getFullyQualifiedName(limeEnumeration),
            limeEnumeration.attributes.have(LimeAttributeType.EXTERNAL_TYPE),
            getPreviousResults(CppEnumItem::class.java)
        )
        cppEnum.comment = limeEnumeration.comment

        storeNamedResult(limeEnumeration, cppEnum)
        closeContext()
    }

    override fun finishBuilding(limeEnumerator: LimeEnumerator) {
        val cppEnumItem = CppEnumItem(
            nameResolver.getName(limeEnumerator),
            nameResolver.getFullyQualifiedName(limeEnumerator),
            getPreviousResultOrNull(CppValue::class.java)
        )
        cppEnumItem.comment = limeEnumerator.comment

        storeNamedResult(limeEnumerator, cppEnumItem)
        closeContext()
    }

    override fun finishBuilding(limeConstant: LimeConstant) {
        val cppConstant = CppConstant(
            nameResolver.getName(limeConstant),
            nameResolver.getFullyQualifiedName(limeConstant),
            getPreviousResult(CppTypeRef::class.java),
            getPreviousResult(CppValue::class.java)
        )
        cppConstant.comment = limeConstant.comment

        storeNamedResult(limeConstant, cppConstant)
        closeContext()
    }

    override fun finishBuilding(limeValue: LimeValue) {
        val valueType = limeValue.typeRef.type
        val isFloat = valueType is LimeBasicType && valueType.typeId == LimeBasicType.TypeId.FLOAT
        val cppValue = when (limeValue) {
            is LimeValue.Literal -> {
                val suffix = if (isFloat) "f" else ""
                CppValue(limeValue.value + suffix)
            }
            is LimeValue.Enumerator ->
                CppValue(nameResolver.getFullyQualifiedName(limeValue.valueRef.enumerator))
            is LimeValue.Special -> {
                val signPrefix = if (limeValue.value == ValueId.NEGATIVE_INFINITY) "-" else ""
                val typeString = if (isFloat) "float" else "double"
                val valueString = if (limeValue.value == ValueId.NAN) "quiet_NaN" else "infinity"
                CppValue(
                    "${signPrefix}std::numeric_limits<$typeString>::$valueString()",
                    listOf(CppLibraryIncludes.LIMITS))
            }
            is LimeValue.Null -> {
                val cppType = getPreviousResult(CppTypeRef::class.java)
                CppValue(
                    "${typeMapper.optionalTypeName}<${cppType.name}>()",
                    listOf(CppLibraryIncludes.OPTIONAL)
                )
            }
            is LimeValue.InitializerList -> CppValue("{}")
        }

        storeResult(cppValue)
        closeContext()
    }

    override fun finishBuilding(limeTypeRef: LimeTypeRef) {
        storeResult(typeMapper.mapType(limeTypeRef))
        closeContext()
    }
}